tests/test_websocket_bootstrap.c

```diff
@@ -0,0 +1,635 @@
+
+static struct aws_websocket_client_bootstrap_function_table s_mock_function_table = {
+    .aws_http_client_connect = s_mock_http_client_connect,
+    .aws_http_connection_release = s_mock_http_connection_release,
+    .aws_http_connection_close = s_mock_http_connection_close,
+    .aws_http_connection_get_channel = s_mock_http_connection_get_channel,
+    .aws_http_stream_new_client_request = s_mock_http_stream_new_client_request,
+    .aws_http_stream_release = s_mock_http_stream_release,
+    .aws_http_stream_get_connection = s_mock_http_stream_get_connection,
+    .aws_http_stream_get_incoming_response_status = s_mock_http_stream_get_incoming_response_status,
+    .aws_websocket_handler_new = s_mock_websocket_handler_new,
+};
+
+static const struct aws_http_header s_handshake_request_headers[] = {
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Host"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("server.example.com"),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Upgrade"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("websocket"),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Connection"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Upgrade"),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Sec-WebSocket-Key"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("dGhlIHNhbXBsZSBub25jZQ=="),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Sec-WebSocket-Version"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("13"),
+    },
+};
+
+static const struct aws_http_header s_handshake_response_headers[] = {
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Upgrade"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("websocket"),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Connection"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Upgrade"),
+    },
+    {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Sec-WebSocket-Accept"),
+        .value = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("s3pPLMBiTxaQ9kYGzzhZRbK+xOo="),
+    },
+};
+
+/* If fail_at_step is set to one of these, that step will explicitly fail and raise its enum value as the error code */
+enum boot_step {
+    BOOT_STEP_HTTP_CONNECT = 0x4000000, /* Use values that don't overlap with another aws-c-xyz library */
+    BOOT_STEP_HTTP_CONNECT_COMPLETE,
+    BOOT_STEP_REQUEST_NEW,
+    BOOT_STEP_REQUEST_COMPLETE,
+    /* If the response validation steps fail, we expect a specific error */
+    BOOT_STEP_VALIDATE_RESPONSE_STATUS = AWS_ERROR_HTTP_WEBSOCKET_UPGRADE_FAILURE,
+    BOOT_STEP_WEBSOCKET_NEW = 0x50000000, /* Back to using made-up error-codes */
+    BOOT_STEP_HTTP_SHUTDOWN,
+};
+
+/* Needs to be a static singleton so that mock functions can access it */
```

Comment:
Isn't it the user data though?

Reply:
Not all the functions being mocked have access to user_data. (ex `aws_http_connection_close(struct aws_http_connection*)`)

but they can access this static singleton